adapters/mediasquare/mediasquare.go

```diff
@@ -0,0 +1,124 @@
+	return &adapter{
+		endpoint: config.Endpoint,
+	}, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var (
+		requestData []*adapters.RequestData
+		errs        []error
+	)
+	if request == nil || request.Imp == nil {
+		errs = append(errs, errorWritter("<MakeRequests> request", nil, true))
+		return nil, errs
+	}
+
+	msqParams := initMsqParams(request)
+	msqParams.Test = (request.Test == int8(1))
+	for _, imp := range request.Imp {
+		var (
+			bidderExt   adapters.ExtImpBidder
+			msqExt      openrtb_ext.ImpExtMediasquare
+			currentCode = msqParametersCodes{
+				AdUnit:    imp.TagID,
+				AuctionId: request.ID,
+				BidId:     imp.ID,
+			}
+		)
+
+		if err := jsonutil.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errs = append(errs, errorWritter("<MakeRequests> imp[ext]", err, len(imp.Ext) == 0))
+			continue
+		}
+		if err := jsonutil.Unmarshal(bidderExt.Bidder, &msqExt); err != nil {
+			errs = append(errs, errorWritter("<MakeRequests> imp-bidder[ext]", err, len(bidderExt.Bidder) == 0))
+			continue
+		}
+		currentCode.Owner = msqExt.Owner
+		currentCode.Code = msqExt.Code
+
+		if ok := currentCode.setContent(imp); ok {
+			msqParams.Codes = append(msqParams.Codes, currentCode)
+		}
+	}
+
+	req, err := a.makeRequest(request, &msqParams)
+	if err != nil {
+		errs = append(errs, err)
+	} else if req != nil {
+		requestData = append(requestData, req)
+	}
+	return requestData, errs
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest, msqParams *msqParameters) (requestData *adapters.RequestData, err error) {
+	var requestJsonBytes []byte
+	if msqParams == nil {
+		err = errorWritter("<makeRequest> msqParams", nil, true)
+		return
+	}
+	if requestJsonBytes, err = jsonutil.Marshal(msqParams); err == nil {
+		var headers http.Header = headerList
+		requestData = &adapters.RequestData{
+			Method:  "POST",
+			Uri:     a.endpoint,
+			Body:    requestJsonBytes,
+			Headers: headers,
+			ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+		}
+	} else {
+		err = errorWritter("<makeRequest> jsonutil.Marshal", err, false)
+	}
+
+	return
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var (
+		bidderResponse *adapters.BidderResponse
+		errs           []error
+	)
+	if response.StatusCode != http.StatusOK {
+		switch response.StatusCode {
+		case http.StatusBadRequest:
+			errs = []error{&errortypes.BadInput{Message: fmt.Sprintf("<MakeBids> Unexpected status code: %d.", response.StatusCode)}}
+		default:
+			errs = []error{&errortypes.BadServerResponse{
+				Message: fmt.Sprintf("<MakeBids> Unexpected status code: %d. Run with request.debug = 1 for more info.", response.StatusCode),
+			}}
+		}
+		return bidderResponse, errs
+	}
```

Comment:
Would you consider to clean up error handling using `CheckResponseStatusCodeForErrors` and `IsResponseStatusCodeNoContent` functions? Please refer to "app nexus.go"->MakeBids. 

I also noticed an interesting thing. When I run the request to mediasquare and bids are not returned - you still return http 200 with error message. This is not typical behavior. If bids are not returned we usually get http 204. If this is something you cannot change, please consider to add a check for empty `responses` after  `jsonutil.Unmarshal(response.Body, &msqResp)`